File: management_app/mock-api-server.py
#!/usr/bin/env python3
"""Mock API Server for WinSLA Frontend Testing"""

from flask import Flask, jsonify, request, make_response
from datetime import datetime

app = Flask(__name__)

# In-memory storage for testing
accounts = []
auth_stats = {
    'connections_accepted': 0,
    'successful_auths': 0,
    'failed_auths': 0
}

@app.route('/api/status', methods=['GET'])
def status():
    auth_stats['connections_accepted'] += 1
    return jsonify({
        'running': True,
        'version': 'v2.2.6',
        **auth_stats
    })

@app.route('/api/accounts', methods=['GET'])
def get_accounts():
    auth_stats['connections_accepted'] += 1
    return jsonify(accounts)

@app.route('/api/accounts', methods=['POST'])
def create_account():
    auth_stats['connections_accepted'] += 1
    data = request.json
    
    new_account = {
        'id': str(datetime.now().timestamp()),
        'account_sid': data.get('account_sid', f'S-1-5-21-{datetime.now().timestamp()}'),
        'account_username': data.get('account_username', 'TestUser'),
        'approvers': [],
        'enabled': True,
        'created_at': datetime.now().isoformat(),
        'updated_at': datetime.now().isoformat()
    }
    accounts.append(new_account)
    return jsonify(new_account), 201

@app.route('/api/accounts/approvers', methods=['POST'])
def add_approver():
    auth_stats['connections_accepted'] += 1
    data = request.json
    
    account_sid = data.get('account_sid')
    approver_sid = data.get('approver_sid')
    approver_username = data.get('approver_username')
    
    account = next((a for a in accounts if a['account_sid'] == account_sid), None)
    if not account:
        return jsonify({'error': 'Account not found'}), 404
    
    # Check if approver already exists
    existing_approver = next((a for a in account['approvers'] if a['sid'] == approver_sid), None)
    if existing_approver:
        return jsonify({'error': 'Approver already exists'}), 409
    
    account['approvers'].append({
        'sid': approver_sid,
        'username': approver_username,
        'enabled': True
    })
    account['updated_at'] = datetime.now().isoformat()
    
    return '', 201

@app.route('/api/accounts/<account_sid>/approvers/<approver_sid>', methods=['DELETE'])
def remove_approver(account_sid, approver_sid):
    auth_stats['connections_accepted'] += 1
    
    account = next((a for a in accounts if a['account_sid'] == account_sid), None)
    if not account:
        return jsonify({'error': 'Account not found'}), 404
    
    initial_length = len(account['approvers'])
    account['approvers'] = [a for a in account['approvers'] if a['sid'] != approver_sid]
    
    if len(account['approvers']) == initial_length:
        return jsonify({'error': 'Approver not found'}), 404
    
    account['updated_at'] = datetime.now().isoformat()
    return '', 204

@app.route('/api/accounts/<account_sid>/enable', methods=['PUT'])
def toggle_account_enabled(account_sid):
    auth_stats['connections_accepted'] += 1
    data = request.json
    
    account = next((a for a in accounts if a['account_sid'] == account_sid), None)
    if not account:
        return jsonify({'error': 'Account not found'}), 404
    
    account['enabled'] = data.get('enabled', False)
    account['updated_at'] = datetime.now().isoformat()
    
    return jsonify({'account_sid': account_sid, 'enabled': account['enabled']})

@app.route('/api/accounts/<account_sid>', methods=['DELETE'])
def delete_account(account_sid):
    auth_stats['connections_accepted'] += 1
    
    initial_length = len(accounts)
    accounts[:] = [a for a in accounts if a['account_sid'] != account_sid]
    
    if len(accounts) == initial_length:
        return jsonify({'error': 'Account not found'}), 404
    
    return '', 204

@app.route('/api/validate-account', methods=['POST'])
def validate_account():
    auth_stats['connections_accepted'] += 1
    data = request.json
    
    # Accept any credentials for testing
    return jsonify({
        'success': True,
        'sid': f'S-1-5-21-mock-{datetime.now().timestamp()}',
        'display_name': data.get('username', 'TestUser'),
        'message': f'验证成功：{data.get("username", "TestUser")}'
    })

if __name__ == '__main__':
    print('=' * 60)
    print('Mock API Server for WinSLA v2.2.6 Testing')
    print('=' * 60)
    print('Available endpoints:')
    print('  GET  /api/status          - Service status')
    print('  GET  /api/accounts        - List all accounts')
    print('  POST /api/accounts        - Create account')
    print('  POST /api/accounts/approvers - Add approver')
    print('  DELETE /api/accounts/:id/approvers/:approverId - Remove approver')
    print('  PUT  /api/accounts/:id/enable     - Toggle enable/disable')
    print('  DELETE /api/accounts/:id      - Delete account')
    print('  POST /api/validate-account  - Validate credentials')
    print('=' * 60)
    print('Server running on http://localhost:19830')
    print('=' * 60)
    app.run(host='0.0.0.0', port=19830, debug=False)

File: management_app/mock-api-server.js
// Mock API Server for Frontend Testing
const express = require('express');
const cors = require('cors');
const path = require('path');

const app = express();
app.use(cors());
app.use(express.json());

// In-memory storage for testing
let accounts = [];
let authStats = {
    connections_accepted: 0,
    successful_auths: 0,
    failed_auths: 0
};

// Status endpoint
app.get('/api/status', (req, res) => {
    authStats.connections_accepted++;
    res.json({
        running: true,
        version: "v2.2.6",
        ...authStats
    });
});

// Get accounts
app.get('/api/accounts', (req, res) => {
    authStats.connections_accepted++;
    res.json(accounts);
});

// Create account
app.post('/api/accounts', (req, res) => {
    authStats.connections_accepted++;
    const newAccount = {
        id: Date.now().toString(),
        account_sid: req.body.account_sid || `S-1-5-21-${Date.now()}`,
        account_username: req.body.account_username || 'TestUser',
        approvers: [],
        enabled: true,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString()
    };
    accounts.push(newAccount);
    res.status(201).json(newAccount);
});

// Add approver
app.post('/api/accounts/approvers', (req, res) => {
    authStats.connections_accepted++;
    const { account_sid, approver_sid, approver_username } = req.body;
    
    const account = accounts.find(a => a.account_sid === account_sid);
    if (!account) {
        return res.status(404).json({ error: 'Account not found' });
    }
    
    // Check if approver already exists
    const existingApprover = account.approvers.find(a => a.sid === approver_sid);
    if (existingApprover) {
        return res.status(409).json({ error: 'Approver already exists' });
    }
    
    account.approvers.push({
        sid: approver_sid,
        username: approver_username,
        enabled: true
    });
    account.updated_at = new Date().toISOString();
    
    res.status(201).send();
});

// Remove approver
app.delete('/api/accounts/:accountSid/approvers/:approverSid', (req, res) => {
    authStats.connections_accepted++;
    const { accountSid, approverSid } = req.params;
    
    const account = accounts.find(a => a.account_sid === accountSid);
    if (!account) {
        return res.status(404).json({ error: 'Account not found' });
    }
    
    const initialLength = account.approvers.length;
    account.approvers = account.approvers.filter(a => a.sid !== approverSid);
    
    if (account.approvers.length === initialLength) {
        return res.status(404).json({ error: 'Approver not found' });
    }
    
    account.updated_at = new Date().toISOString();
    res.status(204).send();
});

// Toggle account enabled
app.put('/api/accounts/:accountSid/enable', (req, res) => {
    authStats.connections_accepted++;
    const { accountSid } = req.params;
    const { enabled } = req.body;
    
    const account = accounts.find(a => a.account_sid === accountSid);
    if (!account) {
        return res.status(404).json({ error: 'Account not found' });
    }
    
    account.enabled = enabled;
    account.updated_at = new Date().toISOString();
    
    res.json({ account_sid: accountSid, enabled });
});

// Delete account
app.delete('/api/accounts/:accountSid', (req, res) => {
    authStats.connections_accepted++;
    const { accountSid } = req.params;
    
    const initialLength = accounts.length;
    accounts = accounts.filter(a => a.account_sid !== accountSid);
    
    if (accounts.length === initialLength) {
        return res.status(404).json({ error: 'Account not found' });
    }
    
    res.status(204).send();
});

// Validate account (mock)
app.post('/api/validate-account', (req, res) => {
    authStats.connections_accepted++;
    const { username, password } = req.body;
    
    // Accept any credentials for testing
    res.json({
        success: true,
        sid: `S-1-5-21-mock-${Date.now()}`,
        display_name: username,
        message: `验证成功：${username}`
    });
});

const PORT = 19830;
app.listen(PORT, () => {
    console.log(`Mock API Server running on http://localhost:${PORT}`);
    console.log('Available endpoints:');
    console.log('  GET  /api/status          - Service status');
    console.log('  GET  /api/accounts        - List all accounts');
    console.log('  POST /api/accounts        - Create account');
    console.log('  POST /api/accounts/approvers - Add approver');
    console.log('  DELETE /api/accounts/:sid/approvers/:approverId - Remove approver');
    console.log('  PUT  /api/accounts/:id/enable     - Toggle enable/disable');
    console.log('  DELETE /api/accounts/:id      - Delete account');
    console.log('  POST /api/validate-account  - Validate credentials');
});
